dom.bs

```diff
@@ -2752,6 +2810,30 @@ before a <var>child</var>, with an optional <i>suppress observers flag</i>, run
  <var>parent</var> with <var>nodes</var>, « », <var>previousSibling</var>, and <var>child</var>.
 
  <li><p>Run the <a>children changed steps</a> for <var>parent</var>.
+
+ <li>
+  <p>Let <var>staticNodeList</var> be a <a for=/>list</a> of <a for=/>nodes</a>, initially
+  empty.</p>
+
+  <p class="note">We collect all <a for=/>nodes</a> <i>before</i> calling the <a>post-insertion
+  steps</a> on any one of them, instead of calling the <a>post-insertion steps</a> <i>while</i>
+  we're traversing the <a>node tree</a>. This is because the <a>post-insertion steps</a> can
+  modify the tree's structure, making live traversal unsafe, possibly leading to the
+  <a>post-insertion steps</a> being called multiple times on the same <a>node</a>.</p>
+ </li>
+
+ <li>
+  <p>For each <var>node</var> in <var>nodes</var>, in <a>tree order</a>:
+
+  <ol>
+   <li><p>For each <a>shadow-including inclusive descendant</a> <var>inclusiveDescendant</var> of
+   <var>node</var>, in <a>shadow-including tree order</a>, <a for=list>append</a>
+   <var>inclusiveDescendant</var> to <var>staticNodeList</var>.
+  </ol>
+ </li>
+
+ <li><p><a for=list>For each</a> <var>node</var> in <var>staticNodeList</var>, if <var>node</var> is
+ <a>connected</a>, then run the <a>post-insertion steps</a> with <var>node</var>.
```

Comment:
Why do we have the requirement for the node to be connected? (It wasn't clear to me either when looking at blink's source code, but I didn't look at the blame for that part).

Reply:
The idea is that post-insertion steps should not run on elements that aren't connected to the DOM. For those elements, we shouldn't be performing actions that modify the frame tree or load resources or anything else the post-insertion steps are responsible for outside of the insertion steps.

We could do this implicitly / vaguely which is kind of how HTML does it now, by specifying something like ["becomes connected"](https://html.spec.whatwg.org/multipage/infrastructure.html#becomes-connected) steps (but for post-insertion steps), which supposedly run "during" the "insertion steps" (without any ordering specified) with a "is connected" precondition built in. Or we could do it slightly more explicitly by not having an "is connected" gate specified in DOM, and relying on each element's post-insertion steps to have a precondition that early-returns without a DOM connection. That doesn't seem better though. So I think the condition in this PR is the most straightforward way of achieving this. We have tests too that ensure e.g., scripts don't run when disconnected after insertion (by other post-insertion steps, in the trickiest of cases, for example).

Reply:
I'd expect the algorithm be generic, and then each post-insertion step implementation to check if the node is connected (if needed). Nothing vague there.

If we go with the approach the pr has, at least the algorithm name should hint that it isn't about generic insertion step, but it is explicitly about inserting to document. Post-connected steps?
(insertion steps algorithm has nothing to do with nodes being connected or not)

Reply:
Hmm, I see the point. For selfish implementer reasons I would gently push back just because both the WebKit and Blink use the word "insertion" in their representations of these steps, so reducing cognitive overhead for implementers seems like a good thing to me. But I am not married to it. If we really wanted to change this, I would!

Reply:
If the semantics really are post-connected in Chromium and Chromium they should just rename. I agree with Olli that we should attempt to have accurate names.

Reply:
I'm assuming you mean Chromium and WebKit? Either way, yes that sounds good to me. Latest commit here renames everything accordingly. I went with `post-connection steps` instead of `post-connected` steps, since `post-connection` is similar to `post-insertion` which is what this PR had before the rename (i.e., we didn't choose `post-inserted steps`). I'm indifferent; if you think it should be renamed to `post-connected` let me know!